skt/publisher.py

```diff
@@ -14,137 +14,74 @@
-    """A copy publisher that copies source to destination."""
-    TYPE = 'cp'
-
-    def publish(self, source):
+        # NOTE(mhayden): scp can handle local file copies properly. Both
+        # 'cp' and 'scp' are left here for backwards compatibility.
+        if self.pub_type in ['cp', 'scp']:
+            self.scp()
+        else:
+            raise(
+                KeyError,
+                "Publisher type not supported: {}".format(self.pub_type)
+            )
+
+        return self.geturl()
+
+    def geturl(self):
         """
-        Copy the source file to public destination.
-
-        Args:
-            source: Source file path.
-
-        Returns:
-            Published URL corresponding to the specified source.
-        """
-        destination = join_with_slash(self.destination, "")
-        shutil.copy(source, destination)
-        return self.geturl(source)
-
-
-class ScpPublisher(Publisher):
-    """A SCP publisher that copies source to (remote) destination."""
-    TYPE = 'scp'
-
-    def publish(self, source):
-        """
-        Copy the source file to public destination.
+        Get published URL for a source file path.
 
         Args:
             source: Source file path.
 
         Returns:
             Published URL corresponding to the specified source.
-        """
-        destination = join_with_slash(self.destination, "")
-        subprocess.check_call(["scp", source, destination])
-        return self.geturl(source)
-
-
-class SftpPublisher(Publisher):
-    """A sftp publisher that copies source to (remote) destination."""
-    TYPE = 'sftp'
-
-    def publish(self, source):
-        """
-        Copy the source file to public destination.
-
-        Args:
-            source: Source file path.
 
-        Returns:
-            Published URL corresponding to the specified source.
         """
-        proc = subprocess.Popen(['sftp', self.destination],
-                                stdin=subprocess.PIPE)
-        proc.stdin.write("put -r %s\n" % source)
-        proc.stdin.close()
-        proc.wait()
-        return self.geturl(source)
-
-
-def getpublisher(ptype, parg, pburl):
-    """
-    Create an instance of a "publisher" subclass with specified arguments.
-
-    Args:
-        rtype:  The value of the class "TYPE" member to match.
-        rarg:   A dictionary with the instance creation arguments.
-
-    Returns:
-        The created class instance.
-
-    Raises:
-        ValueError if the rtype match wasn't found.
-    """
-    for cls in Publisher.__subclasses__():
-        if cls.TYPE == ptype:
-            return cls(parg, pburl)
-    raise ValueError("Unknown publisher type: %s" % ptype)
+        return join_with_slash(self.baseurl, os.path.basename(self.source))
+
+    def scp(self):
+        """Use scp to copy the source to the destination."""
+        # Create the destination directory.
+        if not os.path.isdir(self.destination):
+            os.makedirs(self.destination)
```

Comment:
This doesn't make sense for remote copies :/. Actually `scp` creates the destination if it doesn't exist.

Reply:
I thought so as well, but if I do this:

```
scp -r /tmp/tmpHdZUbM/kernel.tar.gz /tmp/tmpHdZUbM/published/
cp: cannot create regular file '/tmp/tmpHdZUbM/published/': Not a directory
```

There's sadness. ;(

Reply:
Oh I see, it creates just depth=1 :smirk:

Reply:
Well, after all if you remove makedirs, it will have the same behavior when copy remotely or locally